### DISCUSSION TITLE
_TypeError cropping up when converting from .state to .notifier

### DISCUSSION THREAD

**Discussion by mark8044** [rrousselGit/riverpod#1984] [Q&A]
Following instructions in the IDE, I've decided to finally update my `.state` to `.notifier` where I'm being told to

So for example:
```
final showTitleProvider = StateProvider.family<bool, String>((ref, id) => false);
```

```
NotificationListener<ScrollNotification>(
                    onNotification: (scrollNotification) {
                      if ((scrollNotification.metrics.extentBefore > 70) && (ref.read(showTitleProvider(widget.myhash).notifier).state == false))
                      {
                                   ....
                                 ref.read(showTitleProvider(widget.myhash).notifier).state = true;
                      }
...
}
```

After the change, occasionally but not always I've started to get errors as following.  I really don't understand the error or where to begin TBH.   
```
The following _TypeError was thrown while notifying listeners for AnimationController:
type '((StateProviderRef<Y0>) => Y0, {String? name, Family<dynamic>? from, Object? argument, List<ProviderOrFamily>? dependencies, (() => String)? debugGetCreateSourceHash}) => StateProvider<bool>' is not a subtype of type '((StateProviderRef<bool>) => bool, {Object? argument, List<ProviderOrFamily>? dependencies, Family<dynamic> from, String? name}) => StateProvider<bool>' of 'function result'
```

Does anyone have an idea of what the source of this problem is or where I can begin to tackle it?

===

On a side note, you may ask why I'm using a `ref.read` in my `build`.  Well I did try to convert it to `ref.watch` as so: `(ref.watch(showTitleProvider(widget.myhash)) == false)` but then my whole UI started to massively stutter as the user scroll.  I assume the combination of a watch and the notification listener is just too taxing

**Comment by rrousselGit** [comment]
Do you still have the error after a hot-restart?

**Reply by mark8044** [reply]
I believe it does continue

I've rolled back to 2.0.0-dev9 and it doesn't seem to happen with that one, for whatever that's worth

**Reply by rrousselGit** [reply]
Could you share a minimal reproducible example? It's hard for me to help without one

**Reply by mark8044** [reply]
My app has turned into such a large beast its hard to extract the meaningful parts for reproduce code.

Does the error message point you in any direction?  I was thinking it might mean the data in that provider is non existent so the problem is that somehow the data got cleared out due to low memory and cache clearing or something like that?  just a random guess since I've never seen such an error message before

Luckily 2.0.0-dev9 is rock solid for me in regards to that error so for now its ok I can get around the issue.  

But incase that error message has any meaning for you to point me in the right direction I can try to investigate more

Thanks!

**Reply by rrousselGit** [reply]
The error doesn't really help, no.

**Comment by mark8044** [comment]
@rrousselGit 

Ok I think I've found the error.  Im putting this here incase someone else runs into this issue in the future:

I had declared my provider as:
```
final showTitleProvider = StateProvider.family((ref, id) => true);
```

Which had been working fine in 2.0.0-dev9, but occasionally (not always) giving me that error I mentioned above in later versions.  The issue was that I didn't explicitly give the StateProvider its type.  I guess it worked for some reason sometimes, but in certain situations would give a fault.  I changed the provider as follows and everything is good:

```
final showTitleProvider = StateProvider.family<bool,String>((ref, id) => true);
```

In my original post I think I showed the incorrect provider that already had <bool,String> there

**Reply by mark8044** [reply]
I take it back.  Error is back :(

```
======== Exception caught by animation library =====================================================
The following _TypeError was thrown while notifying listeners for AnimationController:
type '((StateProviderRef<Y0>) => Y0, {String? name, Family<dynamic>? from, Object? argument, List<ProviderOrFamily>? dependencies, (() => String)? debugGetCreateSourceHash}) => StateProvider<bool>' is not a subtype of type '((StateProviderRef<bool>) => bool, {Object? argument, List<ProviderOrFamily>? dependencies, Family<dynamic> from, String? name}) => StateProvider<bool>' of 'function result'
```
